### DISCUSSION TITLE
Action hook triggers before creating an item which is in m2m relationship

### DISCUSSION THREAD

**Discussion by MarkoMestrovicShape** [directus/directus#10196] [Archive]
hook action items.create fires before item is being saved to database for m2m items

### To Reproduce

Create an m2m relationship between collections A and B.
Create a hook with  action('B.items.create') which checks the current state of the database with ItemsService and ReadByQuery and the key it got from the context.
Create an item for collection A.
Create an item for B collection from the screen in A collection.
Save changes.
what happens: inside the action callback function when querying ItemService by the key from the context it does not have it.

**Comment by rijkvanzanten** [comment]
Heya! I think this is mostly caused by the order of operations, and the fact that we're working with transactions.

When you're saving a nested relational object, like:

```
articles
categories (o2m to junction)
  articles_id (m2o to parent)
  category_id (m2o to related)
    name: Example
```

The records are inserted in a particular order, in order to resolve the primary keys correctly. In this particular case, the `articles` item would be saved first, as we need the primary key from `articles` to populate `articles_id` in the junction table. Likewise, the new category item will be created before the o2m junction row. If you add a hook on the creation of `articles`, it's expected that the junction may not exist yet.

Next to this, a thing to be aware of is that the whole nested insertion is done in a single transaction. If you use an action hook that fires a secondary query _while_ the first transaction is still in operation, the items may not exist in the database yet, as the transaction isn't done yet. Action/filter hooks get the current database transaction passed in in their context object, which makes it possible to query items that are created in the transaction, before the transaction itself has been committed yet, for example:

```js
module.exports = function registerHook({ filter }, { services }) {
  const { ItemsService } = services;

  filter("B.items.create", (input, meta, { database, schema, accountability }) => {
    const service = new ItemsService('B', { database, schema, accountability });
    // service now uses the current transaction
  });
});
```

**Reply by MarkoMestrovicShape** [reply]
I've put in the parameters of database and accountability and yet the item still doesn't appear when I run the readByQuery and list all of the items in the collection. 

Interestingly enough when I use the axios inside the function, the database updates.

**Reply by rijkvanzanten** [reply]
> Interestingly enough when I use the axios inside the function, the database updates.

When you use axios in the function? You mean when you make a network request to fetch the data?

**Reply by MarkoMestrovicShape** [reply]
No, when I post or get something from an outside website.

Hm, I think the best way to show you would be with showing snippets of code in question.

A and B are in o2m relationship.
I am creating a B item from the A screen.

```js
module.exports = function registerHook({ action }, { services, exceptions }) {
  const { ItemsService } = services;
  const { InvalidPayloadException } = exceptions;

  let checkingFunction = async function (keys, payload, schema, accountability, database) {
    const Bservice = new ItemsService("B", { database, schema, accountability });
    let BItem = await Bservice.readByQuery({
      fields: ["field1", "field2"],
    });
    console.log(BItem); //here, the new item does not appear in the list
    try {
      const response = await axios.post(thirdPartyWebsite, body, { headers });
    } catch (e) {
      let BItem = await Bservice.readByQuery({
        fields: ["field1", "field2"],
      });
      console.log(BItem); //here, the new item from the current transaction can be visible in the database
    }
  };

  action("B.create", (payload, { schema, accountability, database }) => {
    checkingFunction([payload.key], payload.payload, schema, accountability, database);
  });
};
```

**Reply by rijkvanzanten** [reply]
When operating on a nested m2m, when multiple items are created, the hook will fire multiple times. It could be that your hook fires for item 1, while item 2 is still being processed. The timing delay from waiting for the request then means that you check again when both are done

**Reply by gkielwasser** [reply]
Hi @rijkvanzanten ,

I didn't start a new discussion as my issue seems closely related to this one.

Even if it is an "expected behavior" do you know which version has introduced such a change? It seems to be a change in recent Directus versions.

I have also checked in my code some action hook where I was expecting to be able to retrieve the whole item and it is also not working anymore breaking lots of logic.

It seems to be be what you mention upper related to transaction and insertion order. Lacking of e2e tests It is a nightmare to understand where I would have created items this way.

To sumarize could you confirm that doing this we are not sure to be able to retrieve the item ?
```
action('items.create', async function ({
	collection,
	key: itemId,
}, { database, schema, accountability }) {
  // Here we are not sure to be able to retrieve the whole item that has been created
  const { ItemsService } = services;
  const service = new ItemsService(collection, { database, schema, accountability });
  const thisItemMightNotExistYet = await service.readOne(itemId)
}
```

I have taken this assumption since the first hook I have ever whrote with Directus :)

**Reply by joselcvarela** [reply]
Hello @gkielwasser 
As we can see here https://github.com/directus/directus/blob/86b7f7f6847a3edbd0dc68cb0f439f086938ef0e/api/src/services/items.ts#L196-L211

Your example should work. Are you facing some issue using this approach? 🤔

**Reply by gkielwasser** [reply]
@joselcvarela Yes I'm. I'm using v9.4.3 and this code was working "previously" whatever the way the entity is created.
Also if I put a breakpoint to inspect my database at this stage, it is correct my entity is not yet there.

**Reply by gkielwasser** [reply]
@joselcvarela  I will try to explains when it works and when it is not as It seems I wrote some wird stuff.

My schema is : a "craftsman" has multiples "warehouses"
This is working as expected:
```
await directus.items('warehouses').createOne({
                    name,
                    location,
                    craftsman:craftsman.id
                })
```

Creating in the opposite way is not working. The creation hook for the "Warehouses" collection is well trigger, but querying the item as explained, the item doesn't exist already.:

```
const warehousesIds = craftsman.warehouses.map(warehouse => warehouse.id) || []
 await directus.items('craftsmen').updateOne(craftsman.id, {
    warehouses: [...warehousesIds, {
                        name,
                        location,
                    }]
     })
```

**Reply by gkielwasser** [reply]
@joselcvarela Do you think I should create a dedicated discussion for my usecase ? This one is flagged as answered and I'm afraid it will be forgotten :)

**Reply by joselcvarela** [reply]
Sorry @gkielwasser it's been a busy week.
Since this seems an issue to me, the best approach would be to open a new one and if you could, post a short video showing what's happening. 👍

**Reply by gkielwasser** [reply]
@joselcvarela no problem, I will try to sumarize what happen in an issue

**Reply by gkielwasser** [reply]
> Sorry @gkielwasser it's been a busy week. Since this seems an issue to me, the best approach would be to open a new one and if you could, post a short video showing what's happening. 👍

@joselcvarela Here is the issue https://github.com/directus/directus/issues/11172
